hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/utils/SystemContractMethodRegistry.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.contract.impl.exec.utils;
+
+import static java.util.Comparator.comparing;
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hedera.node.app.service.contract.impl.exec.utils.SystemContractMethod.CallVia;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Collection;
+import java.util.SequencedCollection;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * A registry for all the system contract methods - their names, selectors, and signatures.
+ *
+ * Methods are added to this registry when they're defined in the various `FooTranslator` classes,
+ * which, since they're all (or should be) Dagger singletons, is done as the smart contract service
+ * is coming up.  Thus, the completed registry is available once processing starts, and so is ready
+ * for use to enumerate all system contract methods.
+ *
+ * The principal use case for this registry is to be able to generate various per-system-contract-method
+ * metrics.
+ */
+@Singleton
+public class SystemContractMethodRegistry {
+
+    private static final int EXPECTED_SYSTEM_CONTRACT_METHODS_UPPER_BOUND = 250;
+    private final ConcurrentHashMap<String, SystemContractMethod> byName =
+            new ConcurrentHashMap<>(EXPECTED_SYSTEM_CONTRACT_METHODS_UPPER_BOUND);
+
+    // The static constant `SystemContractMethod`s in each of the translator classes do _not_ have
+    // the `SystemContract` field set.  That's set later after all the static fields are created.
+    // So there needs to be a way to get from the version _without_ the system contract to the version
+    // _with_ the system contract.  That's this map.
+    //
+    // Thus, in the running system there are two instances of `SystemContractMethod` for each system
+    // contract method:  One without the `SystemContract`, one with.  This is an acceptable fixed
+    // overhead.  Alternative would have been to bite the bullet and explicitly set the system contract
+    // field in the constructor of each of the static constants in each of the translator classes.
+    // Perhaps that's a refactor that would be worth it sometime.
```

Comment:
I understand the need for this and the refactor you propose seems like a good idea to me.  I do have a concern that the mental effort needed to understand why this is necessary seems to be not small and thus whatever we can do via documentation/comments will be greatly helpful for the future.